### PR TITLE
[Proactive]add support for scheduledtask

### PR DESCRIPTION
**Pull request by lzc850612**
## Description
This change is to add support for scheduledtask, to enable proactive scenarios such as 'Tell me to say hi to mother at 6'. With ScheduledTask support, we can implement scenarios like this.

Changes include:
Added ScheduledTask
Move everything under Utilities to Microsoft.Bot.Solutions
Use TryAdd for turnState operation

## Checklist
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the appropriate unit tests
- [ ] I have tested any new/updated dialogs using Speech in the emulator to ensure the [Speak](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-text-to-speech?view=azure-bot-service-3.0) property is set to enable a high quality speech-first experience and the appropriate [InputHints](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-add-input-hints?view=azure-bot-service-3.0) are set correctly. 
- [ ] I have updated related documentation

If this contains changes that needs to be replicated between the Enterprise Template <-> Virtual Assistant
- [ ] A duplicate issue is filed to track future  work.

If you have updated responses or `.lu` files:
- [ ] All languages have been updated
- [ ] You have tested deployment with your new models

### CODE REVIEW

File: solutions/Virtual-Assistant/src/csharp/VirtualAssistant.sln [hunk 1]
```diff
@@ -1,13 +1,13 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.28010.0
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.28616.25
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/VirtualAssistant.sln]: Was this an intended change or more of the VS odd behaviour? :)

File: solutions/Virtual-Assistant/src/csharp/assistant/Startup.cs [hunk 2]
```diff
@@ -61,8 +61,8 @@ public void ConfigureServices(IServiceCollection services)
             // Initializes your bot service clients and adds a singleton that your Bot can access through dependency injection.
             var languageModels = Configuration.GetSection("languageModels").Get<Dictionary<string, Dictionary<string, string>>>();
             var skills = Configuration.GetSection("skills").Get<List<SkillDefinition>>();
-            var proactiveScenariosConfig = Configuration.GetSection("proactiveSteps").Get<List<ProactiveStep>>();
-            var connectedServices = new BotServices(botConfig, languageModels, skills, proactiveScenariosConfig);
+            var skillEventsConfig = Configuration.GetSection("skillEvents").Get<List<SkillEvent>>();
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/assistant/Startup.cs]: error handling required if the configuration entry isn't found? it's passed into botservices that doesn't do guard checking either.

File: solutions/Virtual-Assistant/src/csharp/assistant/skillEvents.json [hunk 3]
```diff
@@ -1,8 +1,8 @@
 {
-  "proactiveSteps": [
+  "skillEvents": [
     {
       "event": "CarStart",
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/assistant/skillEvents.json]: perhaps rename to DeviceStart to make it more generic than just the auto scenarios?

File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/TaskExtensions/ScheduledProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                    if (!schedule.CancellationToken.IsCancellationRequested)
+                    {
+                        var cronSchedule = CrontabSchedule.Parse(schedule.ScheduleExpression);
+                        var nextrun = cronSchedule.GetNextOccurrence(now);
+                        if (now > nextrun)
+                        {
+                            _backgroundTaskQueue.QueueBackgroundWorkItem(schedule.Task);
+                        }
+
+                        await Task.Delay(100, stoppingToken); // 100 milliseconds delay to next task in line
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/TaskExtensions/ScheduledProcessor.cs]: should we move this to configuration of at the very least const variables at the top?

**Review by darrenj — CHANGES_REQUESTED** [review]
Nothing major, just some quick observations. We should also add a new documentation page around proactive with some high level guidance and we can then add more detail moving forward. Perhaps introduce the architecture, how to configure a simple scenario and example on how to test?

**Comment by lzc850612** [comment]
> Nothing major, just some quick observations. We should also add a new documentation page around proactive with some high level guidance and we can then add more detail moving forward. Perhaps introduce the architecture, how to configure a simple scenario and example on how to test?

yes i inteneded to do this in my next PR.

File: solutions/Virtual-Assistant/src/csharp/VirtualAssistant.sln (hunk 1 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/VirtualAssistant.sln]: no :(. i used visual studio 2019 preview for developing. i'll revert this back.

File: solutions/Virtual-Assistant/src/csharp/assistant/skillEvents.json (hunk 3 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/assistant/skillEvents.json]: good point.

File: solutions/Virtual-Assistant/src/csharp/assistant/Startup.cs (hunk 2 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/assistant/Startup.cs]: good point.

File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/TaskExtensions/ScheduledProcessor.cs (hunk 4 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/TaskExtensions/ScheduledProcessor.cs]: i'll move them to const

**Review by darrenj — APPROVED** [review]

**Pull request merged**